Step3  - 구간 추가 기능

ssamzag opened the pull request:
안녕하세요 김석홍 리뷰어님!
이번 미션은 구현하는데 오래 걸렸네요. 다 끝나고 나니 별 거 아니지만 짤 때는 많이 헤멨네요.. 😅 

그나저나 제가 커밋 단위를 너무 크게 해버린 게 아닌가 싶습니다. 
2단계 미션 리팩토링 후 바로 커밋을 안 하는 바람에 3단계 미션과 같이 커밋을 하게 됐네요. 
feat&refactor 프리픽스를 같이 붙이긴 했는데 이럴 경우 커밋이 가장 애매한 것 같아요.. 좋은 방법이 있을까요?

아무튼 이번 미션도 리뷰 잘 부탁드립니다!!

ssamzag committed:
refactor: 미사용 코드 제거 및 서비스 레이어의 findById 메소드명 변경

ssamzag committed:
docs: 3단계 구간등록 요구사항 및 구현목록 정리

ssamzag committed:
feat&refactor: 2단계 피드백 반영 및 3단계 구간추가 코드 구현&테스트
-

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -5,16 +5,17 @@
 
 @Service
-@Transactional(readOnly = true)
+@Transactional
```
`readOnly` 설정이 빠졌으니, 읽기/쓰기 메소드의 `Transactional` 설정도 같이 확인해보시면 좋을 것 같아요!

ssamzag replied:
개별로 처리해 줘야겠군요! 수정할게요~ 😄

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -43,28 +43,41 @@ public List<LineResponse> findAllLines() {
+
+    private Line findByIdOrElseThrow(Long id) {
+        return lineRepository.findById(id)
+                .orElseThrow(() -> new DataNotFoundException("노선이 존재하지 않습니다."));
```
어떤 아이디로 조회했을 때 발생한 에러인지도 메세지에 포함하면, 예외를 분석할 때 수월할 것 같습니다. 😀

ssamzag replied:
custom 예외를 만들 때 id를 파라미터로 넘겨서 출력하도록 구현하는 게 좋은 방법일까요?

seokhongkim replied:
메세지에 포함시켜도 괜찮을 것 같습니다.
개발자가 보기 위해 예외의 메세지는 인터셉터 등을 활용하여 로그로 남기고, 실제 사용자에게 보여지는 메세지는 다른 메세지로 변경해서 보여주는 방법도 가능할 것 같습니다!

ssamzag replied:
넵 알겠습니다! 좋은 방법 알려 주셔서 감사합니다!!

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,116 @@
+        return distance;
+    }
+
+    public void updateUpStation(Station station, int distance) {
```
로직으로 보니 입력받은 `distance` 가 중간에 추가되는 구간의 거리(e.g., A - B 중간에 C - B 가 추가되면, C - B 사이의 거리)로 보이는데, 맞을까요?
위의 예시에서 A - C 사이의 거리를 계산해서 전달해주지 않고, C - B 사이의 거리를 전달해준 후 내부에서 계산하도록 작성하신 이유가 있을까요?
`Section` 만 봤을 때 `distance` 계산이 왜 이렇게 되는지 이해하기 어려울 것 같아서 궁금합니다.

ssamzag replied:
메시지를 던져서 처리하라는 우테캠 학습권고(?)에 따라 내부에서 처리하도록 구현했습니다. 음.. 이 코드만 다시 보니 사람이 읽기엔 가독성이 떨어지긴하네요.. 😢 
그럼 updateUpStation() 과 updateDistance() 로 나누거나 updateUpStationAndDistance 메소드명을 변경하는 건 별로일까요?

seokhongkim replied:
메시지를 던져서 처리하도록 하신 부분은 좋았습니다. 👍 
위 내용에 대해서는 아래와 같은 방법도 가능할 것 같습니다.
1. distance 도 업데이트해줘야 하는 값을 전달한다.
2. 새로 추가되는 섹션을 전달하고, 이를 바탕으로 기존 섹션의 정보(상행역 or 하행역, 거리)를 업데이트 하도록 요청한다.

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    private void validateDistance(int distance) {
+        if (this.distance < distance) {
```
거리가 같은 경우에도 밸리데이션이 실패해야 할 것 같아요.

ssamzag replied:
넵 추가해 놓을게요!

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,116 @@
+
+    private void validateStation(Station upStation, Station downStation) {
+        if (upStation.getId().equals(downStation.getId())) {
+            throw new IllegalArgumentException("상행역과 하행역은 같을 수 없습니다.");
```
상행역, 하행역을 업데이트 하는 경우에도 기존 역과 변경하는 역이 동일한지 확인할 때 사용하는 메소드 같은데, 예외 메세지가 이 내용을 포함하지는 못하는 것 같습니다.

ssamzag replied:
확인해 보겠습니다!!

seokhongkim commented on src/test/java/nextstep/subway/domain/SectionTest.java:
```diff
@@ -0,0 +1,61 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class SectionTest {
```
꼼꼼한 단위 테스트 작성 좋습니다! 👍

seokhongkim commented on src/test/java/nextstep/subway/domain/StationTest.java:
```diff
@@ -0,0 +1,11 @@
+package nextstep.subway.domain;
+
+class StationTest {
```
테스트에 사용하기 위한 지하철 역을 모아 놓으신 것 같네요.
`test fixture` 라는 개념을 학습해보신 후, 적절한 클래스로 이동해보셔도 좋을 것 같아요!

ssamzag replied:
text fixture 찾아 보겠습니다!
해당 객체가 변경될 일이 없어서 편하게 쓰고자 static으로 만들어 둔건데, 잘못 되었나 보네요 😢

ssamzag replied:
테스트도 기본적인 값들이 세팅되어 있는 factory 패턴을 만들어 두기도 하네요. 이번 단계에서는 팩토리까진 필요없었지만 새로운 개념을 학습하게 되어 매우 좋았습니다!! 👍

seokhongkim replied:
잘못된건 아닙니다 ㅎㅎ 테스트 간에 불필요한 의존성이 생기는 것으로 보여서 학습해보시면 도움이 될 것 같았습니다. 😀

seokhongkim reviewed (changes requested):
민호님, ATDD 3단계 미션 잘 구현해주셨습니다. 👍 

미션의 로직이 갑자기 복잡해져서 헤매셨을 수 있을 것 같아요.
그래도 작은 메소드로 기능을 분리하며 잘 구현해주셨습니다.

말씀하신 것처럼 이번 단계에서는 커밋을 크게, 한 뭉치로 진행하신 것 같네요.
단위 테스트와 일부 기능을 구현한 상태에서 중간 커밋을 나눠서 해보셨어도 좋았을 것 같다는 생각이 듭니다.

그리고 개인적인 생각으로는 한 커밋에서 피처를 구현한 후 리팩토링까지 마무리 하신 것이기 때문에, 리팩토링 프리픽스는 추가하지 않아도 될 것 같습니다. 😀 

코멘트를 조금 남겼으니 확인 부탁드립니다. 🙏

ssamzag committed:
docs: 코드리뷰 피드백 링크 추가

ssamzag committed:
refactor: 3단계 피드백 반영
- 서비스 메소드별 Transactional 설정
- 예외 메시지 상세
- 구간거리 업데이트 메소드명 변경
- 구간거리 유효성체크 변경(동일한 경우에도 예외)

seokhongkim commented on README.md:
```diff
@@ -123,3 +123,74 @@ This project is [MIT](https://github.com/next-step/atdd-subway-admin/blob/master
+  * LineService.java
+    * 구간 추가 및 조회 메서드 추가
+
+#### [코드리뷰 피드백](https://github.com/next-step/atdd-subway-admin/pull/739) (리뷰어: 김재연 님)
```
다른 미션의 리뷰어 이름을 작성해주신 것 같아요 :)

ssamzag replied:
앗 정말 죄송합니다 석홍님 ㅠㅠ 복붙의 위험성이..

seokhongkim reviewed (approved):
민호님, 3단계 미션 피드백 잘 반영해주셨습니다. 👍 

3단계 PR은 이만 머지하도록 하겠습니다.
코멘트에 추가 의견을 조금 남겼습니다. 다음 단계에서 참고하셔도 좋을 것 같습니다.

그럼 4단계 미션에서 뵙겠습니다! 😀
